src/persist-client/src/bin/persist_open_loop_benchmark.rs

```diff
@@ -0,0 +1,449 @@
+
+use anyhow::bail;
+use tokio::sync::Barrier;
+use tokio::task::JoinHandle;
+use tracing::{error, info, trace};
+use tracing_subscriber::EnvFilter;
+
+use mz_persist::workload::DataGenerator;
+use mz_persist_client::Client;
+use mz_persist_client::Location;
+
+// Disable jemalloc on macOS, as it is not well supported [0][1][2].
+// The issues present as runaway latency on load test workloads that are
+// comfortably handled by the macOS system allocator. Consider re-evaluating if
+// jemalloc's macOS support improves.
+//
+// [0]: https://github.com/jemalloc/jemalloc/issues/26
+// [1]: https://github.com/jemalloc/jemalloc/issues/843
+// [2]: https://github.com/jemalloc/jemalloc/issues/1467
+#[cfg(not(target_os = "macos"))]
+#[global_allocator]
+static ALLOC: tikv_jemallocator::Jemalloc = tikv_jemallocator::Jemalloc;
+
+/// Different benchmark configurations.
+#[derive(clap::ArgEnum, Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq)]
+enum BenchmarkType {
+    /// Data is written straight into persistence from the data generator.
+    RawWriter,
+    /// A simulated materialize source pipeline where data is first timestamped using a consensus
+    /// persist shard and then written to persistence.
+    MzSourceModel,
+}
+
+/// Open-loop benchmark for persistence.
+#[derive(clap::Parser)]
+struct Args {
+    /// Number of tokio worker threads.
+    #[clap(short, long, env = "TOKIO_THREADS", value_name = "T")]
+    tokio_worker_threads: Option<usize>,
+
+    /// Number of writer instances.
+    #[clap(long, env = "NUM_WRITERS", value_name = "W", default_value_t = 1)]
+    num_writers: usize,
+
+    /// Number of reader instances.
+    #[clap(long, env = "NUM_READERS", value_name = "R", default_value_t = 1)]
+    num_readers: usize,
+
+    /// Handle to the persist consensus system.
+    #[clap(long, env = "CONSENSUS_URI", value_name = "CONSENSUS_URI")]
+    consensus_uri: String,
+
+    /// Handle to the persist blob storage.
+    #[clap(long, env = "BLOB_URI", value_name = "BLOB_URI")]
+    blob_uri: String,
+
+    /// The type of benchmark to run
+    #[clap(arg_enum, long, required = true)]
+    benchmark_type: BenchmarkType,
+
+    /// Runtime in seconds
+    #[clap(long, value_name = "S", default_value_t = 60)]
+    runtime_seconds: usize,
+
+    /// How many records writers should emit per second.
+    #[clap(long, value_name = "R", default_value_t = 100)]
+    records_per_second: usize,
+
+    /// Size of records (goodbytes) in bytes.
+    #[clap(long, value_name = "B", default_value_t = 64)]
+    record_size_bytes: usize,
+
+    /// Batch size in number of records (if applicable).
+    #[clap(long, env = "", value_name = "R", default_value_t = 10)]
+    batch_size: usize,
+}
+
+const NO_TIMEOUT: Duration = Duration::from_secs(1_000_000);
+// TODO: we currently have a hardcoded 5 second output logging granularity. Perhaps
+// instead we should make that configurable?
+const FIVE_SECONDS: Duration = Duration::from_secs(5);
```

Comment:
because when running with a 2 writer + 2 reader configuration we had 6 tasks emitting log lines (2 datagen + 2 writers + 2 readers) so one line per sec felt really verbose and annoying to read.

Reply:
I think we should be optimizing this for the 1 writer + 1 reader case first. I also think we could get rid of the datagen lines, especially at the levels output by default, that would get us down to 2 lines per second, which feels okay

Reply:
I'll make the logging granularity a command line argument and default it to one second. I'll also keep the datagen lines by default just because i think its a nice to have and we can demote them later as needed.

Reply:
sure, we should definitely keep the datagen lines. let's demote them to debug! or trace! though. it's easy enough to override the log filter on a module basis while you're debugging the tool but the more I think about it, the more they seem noisy for someone just running the benchmark

Reply:
done and done